01_classify_stratified/nbconverted/plot_cancer_type_results.py

```diff
@@ -283,3 +285,136 @@ def label_points(x, y, gene, sig, ax):
+# 
+# We want to see if these confusion matrices lead to reasonable "cancer groups" (i.e. are cancer types that we would expect to be similar close by in the "similarity matrix" given by model confusion?). To do this, we'll just do a super simple agglomerative hierarchical clustering.
+
+# In[14]:
+
+
+# cluster and plot dendrogram
+from scipy.cluster.hierarchy import dendrogram
+from sklearn.cluster import AgglomerativeClustering
+
+# https://github.com/scikit-learn/scikit-learn/blob/70cf4a676caa2d2dad2e3f6e4478d64bcb0506f7/examples/cluster/plot_hierarchical_clustering_dendrogram.py
+def plot_dendrogram(model, **kwargs):
+    
+    # Distances between each pair of children
+    # Since we don't have this information, we can use a uniform one for plotting
+    distance = np.arange(model.children_.shape[0])
+
+    # The number of observations contained in each cluster level
+    no_of_observations = np.arange(2, model.children_.shape[0]+2)
+
+    linkage_matrix = np.column_stack([model.children_,
+                                      distance,
+                                      no_of_observations]).astype(float)
+
+    # Plot the corresponding dendrogram
+    dendrogram(linkage_matrix, **kwargs)
+    
+# fit agglomerative clustering model to confusion (similarity) matrix
+model = AgglomerativeClustering(distance_threshold=0,
+                                n_clusters=None,
+                                affinity='precomputed',
+                                linkage='complete')
+# takes a distance matrix (1 - similarity)
+model = model.fit(1 - preds_expression_df.values)
+
+plot_dendrogram(model, labels=preds_expression_df.index.values)
+plt.title('Dendrogram for complete-linkage hierarchical clustering, expression data',
+          size=14, pad=10)
+plt.xlabel('Cancer type', size=14)
+
+
+# In[15]:
+
+
+# fit agglomerative clustering model to confusion (similarity) matrix
+model = AgglomerativeClustering(distance_threshold=0,
+                                n_clusters=None,
+                                affinity='precomputed',
+                                linkage='complete')
+# takes a distance matrix (1 - similarity)
+model = model.fit(1 - preds_methylation_df.values)
+
+plot_dendrogram(model, labels=preds_methylation_df.index.values)
+plt.title('Dendrogram for complete-linkage hierarchical clustering, methylation data',
+          size=14, pad=10)
+plt.xlabel('Cancer type', size=14)
+
+
+# ## Confusion matrix and clustering with subsampled cancer types
+# 
+# In the expression confusion matrix above, there are some strange "horizontal bars" (e.g. for PAAD and KICH) that we think are an artifact of label imbalance between cancer types of different sizes. To test this, we subsampled each cancer type down to the size of the smallest cancer type (~50 samples) and ran the same classification experiments, taking the average over 5 different random seeds to account for variability in the subsampling.
```

Comment:
Can you elaborate a bit more on how the label imbalance could cause these horizontal lines? So if cancer A had 50 samples vs all other cancers have many more samples, then you think the model would learn to classify all samples as negative . Wouldn't that produce variable performance across different test cancers because different test cancers will have different numbers of samples?

Reply:
The performance might be variable, but what I'm plotting is the probability of the positive class, which I think would be (almost) the same across different test cancers in the case you're describing.

Truthfully I don't think the reason it works has totally clicked for me yet (although your questions are helping me make sense of it!), but in practice the subsampling does seem to fix the problem. Not a very satisfying answer, I know...

Reply:
I think that's how it goes sometimes...sometimes you have to find a solution and then back track why it works:) Something I'm also learning